baseplate/config.py

```diff
@@ -333,30 +332,125 @@ def __init__(self):
         self.__dict__ = self
 
 
-def _parse_config_section(config, spec, root):
-    parsed = ConfigNamespace()
-    for key, parser_or_spec in spec.items():
-        assert "." not in key, "dots are not allowed in keys"
+class Parser(object):
+    """Base for config parsers."""
 
-        if root:
-            key_path = "%s.%s" % (root, key)
-        else:
-            key_path = key
-
-        if callable(parser_or_spec):
-            parser = parser_or_spec
-            raw_value = config.get(key_path, "")
-
-            try:
-                parsed[key] = parser(raw_value)
-            except Exception as e:
-                raise ConfigurationError(key, e)
-        elif isinstance(parser_or_spec, dict):
-            subspec = parser_or_spec
-            parsed[key] = _parse_config_section(config, subspec, root=key_path)
+    @classmethod
+    def from_spec(self, spec):
+        """Return a parser for the given spec object."""
+        if isinstance(spec, Parser):
+            return spec
+        elif isinstance(spec, dict):
+            return SpecParser(spec)
+        elif callable(spec):
+            return CallableParser(spec)
         else:
-            raise AssertionError("invalid specification: %r" % parser_or_spec)
-    return parsed
+            raise AssertionError("invalid specification: %r" % spec)
+
+    def parse(self, key_path, raw_config):
+        """Parse and return the relevant info for a given key.
+
+        :param str key_path: The key this parser is looking for.
+        :param dict raw_config: The full raw configuration dictionary.
+
+        """
+        raise NotImplementedError
+
+
+class SpecParser(Parser):
```

Comment:
Do we document what this looks like anywhere?  Or is a static specification essentially just a non-`DictOf`?